modules/foreman_compute_resource.py

```diff
@@ -0,0 +1,211 @@
+import ansible.module_utils.ansible_nailgun_cement as cement
+
+
+def get_provider_params(provider):
+    provider_name = provider.lower()
+
+    if provider_name == 'libvirt':
+        return {
+            'credentials': ['url'],
+            'params': ['display_type'],
+            'class': nailgun.entities.LibvirtComputeResource
+        }
+
+    elif provider_name == 'ovirt':
+        return {
+            'credentials': ['url', 'user', 'password'],
+            'params': [],
+            'class': cement.OVirtComputeResource
+        }
+
+    elif provider_name == 'vmware':
+        return {
+            'credentials': ['url', 'user', 'password'],
+            'params': ['datacenter'],
+            'class': cement.VMWareComputeResource
+        }
+
+    else:
+        return False
+
+
+def main(module):
+    if not HAS_NAILGUN_PACKAGE:
+        module.fail_json(
+            msg="""Missing required nailgun module (check docs or install)
+            with: pip install nailgun"""
+        )
+
+    name = module.params.get('name')
+    state = module.params.get('state')
+    provider = module.params.get('provider').title()
+    description = module.params.get('description')
+    locations = module.params.get('locations')
+    organizations = module.params.get('organizations')
+
+    cement.create_server(
+        server_url=module.params.get('server_url'),
+        auth=(module.params.get('username'), module.params.get('password')),
+        verify_ssl=module.params.get('verify_ssl'),
+    )
+
+    cement.ping_server(module)
+
+    data = {
+        'name': name,
+        'description': description
+    }
+
+    compute_resource = cement.find_compute_resource(module, name=data.get('name'), failsafe=True)
+
+    if organizations:
+        data['organization'] = [cement.find_organization(module, organization) for organization in organizations]
+
+    if locations:
+        data['location'] = [cement.find_location(module, location) for location in locations]
+
+    data['provider'] = provider
+    provider_params = get_provider_params(provider=provider)
+
+    if not provider_params:
+        module.fail_json(msg='unsupported provider')
+
+    if state != 'latest':
+        for key in provider_params.get('credentials'):
+            data.__setitem__(key, module.params.get('provider_auth').get(key))
+
+    for key in provider_params.get('params'):
+        if key not in module.params:
+            module.fail_json(msg='missing required param {}'.format(key))
+
+        data.__setitem__(key, module.params.get(key))
+
+    changed = cement.naildown_entity_state(provider_params.get('class'), data, compute_resource, state, module)
+
+    return changed
+
+if __name__ == '__main__':
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(type='str', required=True),
+            description=dict(type='str'),
+            provider=dict(type='str', choices=['vmware', 'libvirt', 'ovirt']),
```

Comment:
I think the provider should be `required`, as when you don't pass it, you end up in `module.fail_json(msg='unsupported provider')`

Reply:
I do think differently: To delete a compute_resource, you do not need it. As well for updating an existing one. The only case this is problematic is, when you actually create one. That can only be checked for later in the code. (After not finding the specified compute resource)

Reply:
ah, right!